frontend/src/components/datasources/datasources.tsx

```diff
@@ -0,0 +1,833 @@
+import { sortBy } from "lodash-es";
+import { logNever } from "@/utils/assertNever";
+import { DatabaseLogo } from "@/components/databases/icon";
+import { EngineVariable } from "@/components/databases/engine-variable";
+import type { VariableName } from "@/core/variables/types";
+import { dbDisplayName } from "@/components/databases/display";
+import { AddDatabaseDialog } from "../editor/database/add-database-form";
+import {
+  tablePreviewsAtom,
+  dataConnectionsMapAtom,
+  DEFAULT_ENGINE,
+} from "@/core/datasets/data-source-connections";
+import { PythonIcon } from "../editor/cell/code/icons";
+import { PreviewSQLTable } from "@/core/functions/FunctionRegistry";
+import { useAsyncData } from "@/hooks/useAsyncData";
+import { DatasourceLabel, EmptyState, RotatingChevron } from "./components";
+
+const sortedTablesAtom = atom((get) => {
+  const tables = get(datasetTablesAtom);
+  const variables = get(variablesAtom);
+  const cellIds = get(cellIdsAtom);
+
+  // Sort tables by the index of the variable they are defined in
+  return sortBy(tables, (table) => {
+    // Put at the top
+    if (!table.variable_name) {
+      return -1;
+    }
+    const variable = Object.values(variables).find(
+      (v) => v.name === table.variable_name,
+    );
+    if (!variable) {
+      return 0;
+    }
+
+    const index = cellIds.inOrderIds.indexOf(variable.declaredBy[0]);
+    if (index === -1) {
+      return 0;
+    }
+    return index;
+  });
+});
+
+const connectionsAtom = atom((get) => {
+  const dataConnections = new Map(get(dataConnectionsMapAtom));
+  const defaultEngine = dataConnections.get(DEFAULT_ENGINE);
+
+  // Filter out the internal duckdb engine if it has no databases
+  if (defaultEngine && defaultEngine.databases.length === 0) {
+    dataConnections.delete(DEFAULT_ENGINE);
+  }
+
+  // Sort by name, but put the default engine at the top
+  return sortBy([...dataConnections.values()], (connection) =>
+    connection.name === DEFAULT_ENGINE ? "" : connection.name,
+  );
+});
+
+export const DataSources: React.FC = () => {
+  const [searchValue, setSearchValue] = React.useState<string>("");
+
+  const closeAllColumns = useSetAtom(closeAllColumnsAtom);
+  const tables = useAtomValue(sortedTablesAtom);
+  const dataConnections = useAtomValue(connectionsAtom);
+
+  if (tables.length === 0 && dataConnections.length === 0) {
+    return (
+      <PanelEmptyState
+        title="No tables found"
+        description="Any datasets/dataframes in the global scope will be shown here."
+        action={
+          <AddDatabaseDialog>
+            <Button variant="outline" size="sm">
+              Add database
+              <PlusIcon className="h-4 w-4 ml-2" />
+            </Button>
+          </AddDatabaseDialog>
+        }
+        icon={<DatabaseIcon />}
+      />
+    );
+  }
+
+  const hasSearch = !!searchValue.trim();
+
+  return (
+    <Command
+      className="border-b bg-background rounded-none h-full pb-10 overflow-auto"
```

Comment:
i noticed a blue border (maybe always was there). we can remove that with `outline-none`